serve/mlc_serve/model/sampler.py

```diff
@@ -0,0 +1,441 @@
+        sampling_tensors.past_output_tokens,
+        sampling_tensors.logit_bias_indices,
+        sampling_tensors.logit_bias_values,
+    )
+
+    # TODO(vvchernov): need to strictly define order of using penalties and logit bias or
+    # prohibit simultaneous using of them. At the latter case it can be LogitProcessor
+    if apply_penalty:
+        repetition_penalties_t = repetition_penalties_t[:, None].repeat(1, vocab_size)
+        logits = torch.where(
+            logits > 0, logits / repetition_penalties_t, logits * repetition_penalties_t
+        )
+        bin_counts = torch.zeros(
+            (batch_size, vocab_size + 1), dtype=torch.long, device=logits.device
+        )
+        bin_counts.scatter_add_(
+            1, past_output_tokens_t, torch.ones_like(past_output_tokens_t)
+        )
+        bin_counts = bin_counts[:, :vocab_size]
+        mask = bin_counts > 0
+        logits -= frequency_penalties_t.unsqueeze_(dim=1) * bin_counts
+        logits -= presence_penalties_t.unsqueeze_(dim=1) * mask
+
+    # Adjust temperature
+    logits.div_(temp_t.unsqueeze(dim=1))
+    if apply_top_p_top_k:
+        logits = _apply_top_p_top_k(logits, top_ps_t, top_ks_t)
+
+    if apply_bias:
+        # logits.scatter_add_ performs the following computation:
+        #   logit[i][index[i][j]] += src[i][j]
+        #     where 0<=i<src.shape[0] && 0<=j<src.shape[1]
+        logits.scatter_add_(dim=1, index=logit_bias_indices_t, src=logit_bias_values_t)
+    return logits
+
+
+@dataclass
+class SamplingOutput:
+    next_tokens: list[int]
+    logprob_infos: list[Optional[RawLogprobsInfo]]
+
+
+def sample(
+    logits: torch.Tensor,
+    sampling_metadata: SamplingMetadata,
+    check_safety: bool = False,
+) -> SamplingOutput:
+    def _is_safe_to_sample(prob_like):
+        return (
+            torch.sum(torch.isnan(prob_like) | torch.isinf(prob_like) | (prob_like < 0))
+            == 0
+        )
+
+    res_greedy, res_random = None, None
+    sampling_tensors = sampling_metadata.sampling_tensors
+
+    batch_size = logits.shape[0]
+    mask_greedy_t, mask_random_t = (
+        sampling_tensors.mask_greedy,
+        sampling_tensors.mask_random,
+    )
+
+    next_tokens = np.empty((batch_size,), dtype=np.int64)
+    if sampling_metadata.has_greedy:
+        res_greedy = torch.argmax(logits[mask_greedy_t], -1)
+        np_mask_greedy = mask_greedy_t.cpu().numpy()
+        next_tokens[np_mask_greedy] = res_greedy.cpu().numpy()
+
+    probs_random = None
+    if sampling_metadata.has_random:
+        probs_random = torch.softmax(logits[mask_random_t], dim=-1)
+        if check_safety and not _is_safe_to_sample(probs_random):
+            return None
+        res_random = torch.multinomial(probs_random, 1, True)[:, 0]
+        np_mask_random = mask_random_t.cpu().numpy()
+        next_tokens[np_mask_random] = res_random.cpu().numpy()
+
+    logprob_infos: List[Optional[RawLogprobsInfo]] = np.full(
```

Comment:
Should it be `[None] * batch_size`?

Reply:
Thanks! I was experimenting with numpy and forgot to revert it back. Now it is reverted.